cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DocumentStore.java

```diff
@@ -0,0 +1,138 @@
+
+    /* This map should only be accessed inside synchronized(openDatastores) blocks */
+    private static final Map<File, DocumentStore> documentStores = new HashMap<File, DocumentStore>();
+
+    private static EventBus eventBus = new EventBus();
+
+    private DocumentStore(File location, KeyProvider keyProvider) throws DatastoreException, IOException, SQLException {
+        this.location = location;
+        this.databaseName = location.getName();
+        this.database = new DatabaseImpl(location, keyProvider);
+        // this will be used to observe purge() events from database and forward them to indexmanager
+        this.database.getEventBus().register(this);
+        this.query = new IndexManagerImpl(database);
+    }
+
+    public static DocumentStore getInstance(File location) throws DatastoreNotCreatedException {
+        return getInstance(location, new NullKeyProvider());
+    }
+
+    public static DocumentStore getInstance(File location, KeyProvider provider) throws DatastoreNotCreatedException {
+        boolean created = checkPathAndCreateIfNeeded(location);
+        try {
+            synchronized (documentStores) {
+                DocumentStore ds = documentStores.get(location);
+                if (ds == null) {
+                    ds = new DocumentStore(location, provider);
+                    documentStores.put(location, ds);
+                    eventBus.post(new DatabaseOpened(ds.databaseName));
+                }
+                if (created) {
+                    eventBus.post(new DatabaseCreated(ds.databaseName));
+                }
+                return ds;
+            }
+        }
+        catch (IOException e) {
+            throw new DatastoreNotCreatedException("Database not found: " + location, e);
+        } catch (SQLException e) {
+            throw new DatastoreNotCreatedException("Database not initialized correctly: " + location, e);
+        } catch (DatastoreException e) {
+            throw new DatastoreNotCreatedException("Datastore not initialized correctly: " + location, e);
+        }
+    }
+
+    private static boolean checkPathAndCreateIfNeeded(File location) {
+        //logger.fine("Datastore path: " + directoryPath);
+
+        boolean created = false;
+        if(!location.exists()){
+            created = location.mkdir();
+        }
+
+        Misc.checkArgument(location.isDirectory(), "Input path is not a valid directory");
+        Misc.checkArgument(location.canWrite(), "Datastore directory is not writable");
+        return created;
+    }
+
+    public void close() {
+        synchronized (documentStores) {
+            DocumentStore ds = documentStores.get(location);
+            if (ds != null) {
+                System.out.println("removing "+location);
+                ((DatabaseImpl)database).close();
+                ((IndexManagerImpl)query).close();
+                documentStores.remove(location);
+            }
+            // TODO warn if not in map?
```

Comment:
If this isn't in the map, it has already been closed, I think we should error out somehow to give a hint to the developer that they should only close this once.